src/conn_user.c

```diff
@@ -0,0 +1,189 @@
+		JLOG_ERROR("agent->config.concurrency_mode=%d Only JUICE_CONCURRENCY_MODE_USER (%d) is supported", 
+		            agent->config.concurrency_mode, JUICE_CONCURRENCY_MODE_USER);
+		mutex_unlock(&conn_impl->mutex);
+		return JUICE_ERR_INVALID;
+	}
+
+	addr_record_t src;
+	int ret = conn_user_recv(conn_impl->sock, buffer, size, &src);
+
+	if (ret < 0) {
+		agent_conn_fail(agent);
+		conn_impl->state = CONN_STATE_FINISHED;
+		mutex_unlock(&conn_impl->mutex);
+		return JUICE_ERR_FAILED;
+	} else if (ret > 0) {
+		if (agent_conn_recv(agent, buffer, (size_t)ret, &src) != 0) {
+			JLOG_WARN("Agent receive failed");
+			conn_impl->state = CONN_STATE_FINISHED;
+			mutex_unlock(&conn_impl->mutex);
+			return JUICE_ERR_FAILED;
+		}
+	}
+
+	if (   ret > 0 // We just received a datagram
+	    || conn_impl->next_timestamp <= current_timestamp()
+	    || agent->state != JUICE_STATE_COMPLETED) {
+		if (agent_conn_update(agent, &conn_impl->next_timestamp) != 0) {
+			JLOG_WARN("Agent update failed");
+			conn_impl->state = CONN_STATE_FINISHED;
+			mutex_unlock(&conn_impl->mutex);
+			return JUICE_ERR_FAILED;
+		}
+	}
+
+	mutex_unlock(&conn_impl->mutex);
+	return ret;
+}
+
+static inline int conn_user_recv(socket_t sock, char *buffer, size_t size, addr_record_t *src) {
+	JLOG_VERBOSE("Receiving datagram");
+	int len;
+	while ((len = udp_recvfrom(sock, buffer, size, src)) == 0) {
+		// Empty datagram, ignore
+	}
+
+	if (len < 0) {
+		if (sockerrno == SEAGAIN || sockerrno == SEWOULDBLOCK) {
+			JLOG_VERBOSE("No more datagrams to receive");
+			return 0;
+		}
+		JLOG_ERROR("recvfrom failed, errno=%d", sockerrno);
+		return -1;
+	}
+
+	addr_unmap_inet6_v4mapped((struct sockaddr *)&src->addr, &src->len);
+	return len; // len > 0
+}
+
+int conn_user_init(juice_agent_t *agent, conn_registry_t *registry, udp_socket_config_t *config) {
+	(void)registry;
+
+	conn_impl_t *conn_impl = calloc(1, sizeof(conn_impl_t));
+	if (!conn_impl) {
+		JLOG_FATAL("Memory allocation failed for connection impl");
+		return -1;
+	}
+
+	conn_impl->sock = udp_create_socket(config);
+	if (conn_impl->sock == INVALID_SOCKET) {
+		JLOG_ERROR("UDP socket creation failed");
+		free(conn_impl);
+		return -1;
+	}
+
+	mutex_init(&conn_impl->mutex, 0);
+	mutex_init(&conn_impl->send_mutex, 0);
+
+	agent->conn_impl = conn_impl;
+
+	return JUICE_ERR_SUCCESS;
+}
+
+void conn_user_cleanup(juice_agent_t *agent) {
+	conn_impl_t *conn_impl = agent->conn_impl;
+
+	closesocket(conn_impl->sock);
+	mutex_destroy(&conn_impl->mutex);
+	mutex_destroy(&conn_impl->send_mutex);
+	free(agent->conn_impl);
+	agent->conn_impl = NULL;
+}
+
+void conn_user_lock(juice_agent_t *agent) {
+	conn_impl_t *conn_impl = agent->conn_impl;
+	mutex_lock(&conn_impl->mutex);
+}
+
+void conn_user_unlock(juice_agent_t *agent) {
+	conn_impl_t *conn_impl = agent->conn_impl;
+	mutex_unlock(&conn_impl->mutex);
+}
+
+int conn_user_interrupt(juice_agent_t *agent) {
+	// juice_user_poll does not block when polling, so there's nothing to interrupt

```

Comment:
You still need to schedule a call to `agent_conn_update()` on the next poll, it could be achieved by setting resetting `next_timestamp` to 0. This is why you currently need to add the `state != JUICE_STATE_COMPLETED` in `juice_user_poll()`.